test_suite/tests/derive.rs

```diff
@@ -233,6 +235,69 @@ fn associated_types_derive_without_bounds() {
     assert_type!(Assoc<ConcreteTypes>, struct_type);
 }
 
+#[test]
+fn scale_compact_types_work_in_structs() {
+    #[allow(unused)]
+    #[derive(Encode, TypeInfo)]
+    struct Dense {
+        a: u8,
+        #[codec(compact)]
+        b: u16,
+    }
+
+    let ty = Type::builder()
+        .path(Path::new("Dense", "derive"))
+        .composite(
+            Fields::named()
+                .field_of::<u8>("a", "u8")
+                .field_of::<Compact<u16>>("b", "u16"),
+        );
+
+    assert_type!(Dense, ty);
+
+    // TODO: These two tests both pass. This illustrates a doubt I have that
+    // this is the right way to go. With this PR we take `#[codec(compact)]` to
+    // mean that the type is encoded/decoded as `Compact<T>` but in reality the
+    // output of the `Encode` macro is something quite different and not
+    // `Compact<T>`.
+    // Maybe this is fine, maybe it's not, just not sure.
```

Comment:
@ascjones I left this TODO as an illustration of the crux of the PR: we will take `#[codec(compact)] thing: T` to mean `thing: Compact<T>`. I am not 100% positive this is correct so I wanted to point it out.

Reply:
I think it is okay, since I believe both these will result in the exact same encoded representation (@thiolliere?), which is what we are describing here. 

I presume one should usually prefer the attribute rather than the type for fields because that is more ergonomic for usage of the type, we shouldn't care about how it is encoded when using it.

Reply:
yes, the macro will use the type: `<$FIELD as HasCompact>::Type`
The trait HasCompact is implemented automatically for `T` and associated type to be `Compact<T>`.
But actually a user of parity-scale-codec can implemement HasCompact himself on its type and have something different that `Compact<T>` used when encoding/decoding.

One solution could be to Seal the trait in parity-scale-codec in order to prevent user to do that.

But I don't know about the architecture of scale-info, but it is possible to actually use the correct type: `<$FIELD as HasCompact>::Type` ? (which will then resolve to `Compact<T>` most of the time)

Reply:
Hmm. I think I'm running into the limits of my brain power here. When I add a `HasCompact` bound [here](https://github.com/paritytech/scale-info/blob/5649a666bb3e8228ebdf59e5844ddf22604594a5/src/ty/fields.rs#L136) and try to `MetaType::new::<<T as HasCompact>::Type>()` I end up with infinite recursion:

```
error[E0275]: overflow evaluating the requirement `Compact<_>: Decode`
  |
  = help: consider adding a `#![recursion_limit="256"]` attribute to your crate (`scale_info`)
  = note: required because of the requirements on the impl of `Decode` for `Compact<_>`
  = note: 126 redundant requirements hidden
  = note: required because of the requirements on the impl of `Decode` for `Compact<_>`
  ```

(it's infinite; the error message increases the suggestion indefinitely, 256, 512, …, … 8192 etc)

Reply:
I guess `<T as HasCompact>::Type` must be `Self`?

Reply:
Or some other way ends up calling back into this `TypeInfo` impl

Reply:
this compiles to me:
```diff
diff --git a/src/build.rs b/src/build.rs
index 9669aee..d288320 100644
--- a/src/build.rs
+++ b/src/build.rs
@@ -263,7 +263,8 @@ impl FieldsBuilder<NamedFields> {
     /// Add a named, [`Compact`] field of type `T`.
     pub fn compact_of<T>(mut self, name: &'static str, type_name: &'static str) -> Self
     where
-        T: TypeInfo + 'static,
+               T: scale::HasCompact,
+               <T as scale::HasCompact>::Type: TypeInfo + 'static,
     {
         self.fields
             .push(Field::compact_of::<T>(Some(name), type_name));
@@ -284,7 +285,8 @@ impl FieldsBuilder<UnnamedFields> {
     /// Add an unnamed, [`Compact`] field of type `T`.
     pub fn compact_of<T>(mut self, type_name: &'static str) -> Self
     where
-        T: TypeInfo + 'static,
+               T: scale::HasCompact,
+               <T as scale::HasCompact>::Type: TypeInfo + 'static,
     {
         self.fields.push(Field::compact_of::<T>(None, type_name));
         self
diff --git a/src/ty/fields.rs b/src/ty/fields.rs
index 9229ab9..974fa09 100644
--- a/src/ty/fields.rs
+++ b/src/ty/fields.rs
@@ -133,9 +133,10 @@ impl Field {
     /// Creates a new [`Compact`] field.
     pub fn compact_of<T>(name: Option<&'static str>, type_name: &'static str) -> Field
     where
-        T: TypeInfo + 'static,
+               T: scale::HasCompact,
+               <T as scale::HasCompact>::Type: TypeInfo + 'static,
     {
-        Self::new(name, MetaType::new::<scale::Compact<T>>(), type_name)
+        Self::new(name, MetaType::new::<<T as scale::HasCompact>::Type>(), type_name)
     }
 }
 
```

I see rustc fails to give an ok error message when src/build bounds are not changed

Reply:
> I see rustc fails to give an ok error message when src/build bounds are not changed

Yeah was wondering about that, it didn't complain. :/ Thank you though, the above type checks for me, but this code still causes infinite recursion:

```rust
Fields::named().compact_of::<u16>("b", "u16"),
```
Full test [here](https://github.com/paritytech/scale-info/blob/5649a666bb3e8228ebdf59e5844ddf22604594a5/test_suite/tests/derive.rs#L261-L284), run with `cargo t --all`.

@ascjones do you mean `type Identity = <Self as HasCompact>::Type;` [here](https://github.com/paritytech/scale-info/blob/5649a666bb3e8228ebdf59e5844ddf22604594a5/src/impls.rs#L260)? In itself it doesn't compile. :/

Reply:
I think it is because derive macro doesn't bound the correct bound on the compact type. (see DM)

Reply:
This is now sorted thanks to @thiolliere. 💯